controllers/k8ssandra/telemetry_reconciler.go

```diff
@@ -0,0 +1,86 @@
+// Logic in this file reconciles telemetry resources into the state declared in the CRs.
+
+package k8ssandra
+
+import (
+	"context"
+	"errors"
+
+	"github.com/go-logr/logr"
+	cassdcapi "github.com/k8ssandra/cass-operator/apis/cassandra/v1beta1"
+	api "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	telemetryapi "github.com/k8ssandra/k8ssandra-operator/apis/telemetry/v1alpha1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/telemetry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *K8ssandraClusterReconciler) reconcileCassandraDCTelemetry(
+	ctx context.Context,
+	kc *api.K8ssandraCluster,
+	dcTemplate api.CassandraDatacenterTemplate,
+	actualDc *cassdcapi.CassandraDatacenter,
+	logger logr.Logger,
+	remoteClient client.Client,
+) (ctrl.Result, error) {
+	logger.Info("reconciling telemetry", "cluster", kc.Name, "datacenter", actualDc.Name)
+	mergedSpec := kc.Spec.Cassandra.CassandraTelemetry.Merge(dcTemplate.CassandraTelemetry)
+	dcCfg := telemetry.CassTelemetryResourcer{
+		CassandraNamespace: kc.Namespace,
+		DataCenterName:     actualDc.Name,
+		ClusterName:        kc.Name,
+		Logger:             logger,
+	}
+	logger.Info("merged TelemetrySpec constructed", "mergedSpec", mergedSpec, "cluster", kc.Name, "datacenter", actualDc.Name)
+	// Confirm telemetry config is valid (e.g. Prometheus is installed if it is requested.)
+	validConfig, err := mergedSpec.IsValid(remoteClient, logger)
+	if err != nil {
+		return ctrl.Result{}, errors.New("could not determine if telemetry config is valid")
+	}
+	if !validConfig {
+		return ctrl.Result{}, errors.New("Telemetry spec was invalid for this cluster - is Prometheus installed if you have requested it?")
+	}
+	// If Prometheus not installed bail here.
+	promInstalled, err := telemetryapi.IsPromInstalled(remoteClient, logger)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if !promInstalled {
+		return ctrl.Result{}, nil
+	}
+	// Determine if we want a cleanup or a resource update.
+	switch {
+	case mergedSpec == nil:
```

Comment:
The first 3 case branches are fairly identical, and similar to the default branch. Is it really useful to distinguish all of them?

Reply:
It is, because if we simply wrote `case mergedSpec.Prometheus.Enabled:` but the `Prometheus` part of the path was nil, we'd get a segfault AGAIK. 

I have been agonising over this however, and I think that one simplification we can make is to make `mergedSpec.Prometheus.Enabled` a bool instead of a bool pointer. That way it is not optional and is always defined when `Prometheus` is non-nil. This eliminates one branch from this switch statement.